apps/api/migrations/expire-at/expire-at.migration.ts

```diff
@@ -0,0 +1,94 @@
+import {
+  MessageRepository,
+  NotificationRepository,
+  ExecutionDetailsRepository,
+  JobRepository,
+  OrganizationRepository,
+  EnvironmentRepository,
+  JobStatusEnum,
+} from '@novu/dal';
+import { addMonths } from 'date-fns';
+
+const messageRepository = new MessageRepository();
+const notificationRepository = new NotificationRepository();
+const jobRepository = new JobRepository();
+const executionDetailsRepository = new ExecutionDetailsRepository();
+const organizationRepository = new OrganizationRepository();
+const environmentRepository = new EnvironmentRepository();
+const now = Date.now();
+const expireAtOneMonth = addMonths(now, 1);
+const expireAtSixMonths = addMonths(now, 6);
+
+export async function createExpireAt() {
+  // eslint-disable-next-line no-console
+  console.log('start migration - add expireAt field');
+
+  // eslint-disable-next-line no-console
+  console.log('get organizations and its environments');
+
+  const organizations = await organizationRepository.find({});
+  for (const organization of organizations) {
+    const environments = await environmentRepository.findOrganizationEnvironments(organization._id);
+    for (const environment of environments) {
+      const query = {
+        _organizationId: organization._id,
+        _environmentId: environment._id,
+        expireAt: { $exists: false },
+      };
+      await messagesSetExpireAt(query);
+      await notificationExpireAt(query);
+    }
+  }
+
+  // eslint-disable-next-line no-console
+  console.log('end migration');
+}
+
+export async function messagesSetExpireAt(query) {
+  await messageRepository.update(
+    {
+      ...query,
+      channel: { $ne: 'in_app' },
+    },
+    { $set: { expireAt: expireAtOneMonth } }
+  );
+
+  await messageRepository.update(
+    {
+      ...query,
+      channel: 'in_app',
+    },
+    { $set: { expireAt: expireAtSixMonths } }
+  );
+}
+
+export async function notificationExpireAt(query) {
+  const excludedIds = await getExcludedNotificationIds(query);
+
+  await notificationRepository.update(
+    { ...query, _id: { $nin: excludedIds } },
+    { $set: { expireAt: expireAtOneMonth } }
+  );
+  await jobRepository.update(
+    { ...query, _notificationId: { $nin: excludedIds } },
+    { $set: { expireAt: expireAtOneMonth } }
+  );
+  await executionDetailsRepository.update(
+    { ...query, _notificationId: { $nin: excludedIds } },
+    { $set: { expireAt: expireAtOneMonth } }
+  );
+}
+
+export async function getExcludedNotificationIds(query) {
+  const pendingNotifications = await jobRepository._model.distinct('_notificationId', {
+    ...query,
+    status: JobStatusEnum.PENDING,
+  });
+
+  // digested events stays pending, leaving the notification and deleting the actually digested could cause errors
+  return await notificationRepository._model.distinct('_id', {
+    ...query,
+    _id: { $in: pendingNotifications },
+    _digestedNotificationId: { $exists: false, $eq: null },
+  });
+}
```

Comment:
so we exclude all the notifications that were created without the digest... is that correct?
when the digest is created for the first time, the notification won't have this field `_digestedNotificationId` set... so we also exclude these notifications?

can you maybe explain this a little? thanks

Reply:
@LetItRock you are right. But, for notifications that have been digested - all jobs for the steps after the digest stays pending (we never change their status) while for the first digest notification all jobs are run to completion. So, first digest will not be excluded in lines 83-36, and those that have been digested will not be excluded by lines 89-93. Meaning, all those cases expireAt will be added for them.